src/analysis/max_min_driving_force.jl

```diff
@@ -55,56 +55,95 @@ optimization problems.
+    metabolite_gibbs_free_energy_error_fallback_fraction = 0.01,
     modifications = [],
+    negligible_flux = 1e-6,
 )
-    # get ΔGs in order of reactions
-    dg0s = [gibbs_free_energies[rid] for rid in reactions(model)]
 
-    # get all metabolites that participate in the ΔG calculations (no protons, no water)
-    midxs = [
-        midx for
-        (midx, mid) in enumerate(metabolites(model)) if !(mid in ignore_metabolites)
-    ] # idx in original model
-    mids = metabolites(model)[midxs]
-
-    # find the corresponding metabolites
-    S = stoichiometry(model)[midxs, :]
-
-    # start building the optimization model
     opt_model = Model(optimizer)
+
     @variables opt_model begin
         minDF
-        logcs[1:length(midxs)]
-        dgs[1:n_reactions(model)]
+        logcs[1:n_metabolites(model)]
+        dgfs[1:n_metabolites(model)]
+        dgrs[1:n_reactions(model)]
     end
 
-    @constraints opt_model begin
-        minDF .<= -dgs
-        dgs .<= 0
-        dgs .== dg0s .+ R * T .* S' * logcs
+    # get metabolite ΔG⁰ of formation in order of metabolites in the model
+    dgf0s = [metabolite_gibbs_free_energies[mid] for mid in metabolites(model)]
+
+    # set proton ΔG formation to zero (biochemical thermodynamic assumption)
+    proton_idxs = Int.(indexin(proton_ids, metabolites(model)))
+    for idx in proton_idxs
+        JuMP.fix(dgfs[idx], 0.0)
+        JuMP.fix(logcs[idx], 0.0)
     end
 
-    @objective(opt_model, Max, minDF)
+    # set water concentrations to zero (aqueous condition assumptions)
+    water_idxs = Int.(indexin(water_ids, metabolites(model)))
+    for idx in water_idxs
+        JuMP.fix(logcs[idx], 0.0)
+    end
+
+    # set bounds relative to supplied data, fallback to 1% of value
+    dgf_errors = [
+        get(
+            metabolite_gibbs_free_energy_errors,
+            mid,
+            metabolite_gibbs_free_energy_error_fallback_fraction *
+            metabolite_gibbs_free_energies[mid],
+        ) for mid in metabolites(model)
+    ]
+    for midx in eachindex(metabolites(model))
+        midx in water_idxs && continue
+        midx in proton_idxs && continue
+        JuMP.set_lower_bound(dgfs[midx], dgf0s[midx] - dgf_errors[midx])
+        JuMP.set_upper_bound(dgfs[midx], dgf0s[midx] + dgf_errors[midx])
+    end
+
+    dgr_adjs =
+        [get(reaction_gibbs_free_energy_adjustments, rid, 0.0) for rid in reactions(model)]
 
-    # add the absolute bounds
-    for (midx, mid) in enumerate(mids) # idx in opt_model (missing ignore_metabolites)
+    S = stoichiometry(model)
+
+    @constraint(opt_model, dgrs .== dgr_adjs .+ S' * (dgfs .+ (R * T) .* logcs))
+
+    # thermodynamics should correspond to the fluxes
+    fluxes = [get(flux_solution, rid, 1.0) for rid in reactions(model)]
+
+    # only consider physiological reactions
+    active_ridxs = [
+        idx for (idx, rid) in enumerate(reactions(model)) if
+        rid ∉ ignore_reactions && negligible_flux <= abs(fluxes[idx])
```

Comment:
𝖚𝖓𝖎𝖈𝖔𝖉𝖊

Reply:
ha ha ha it is just such a pleasing operator :)

Reply:
btw (readability hint) in comparisons, can we put the "active" thing to the left? this looks like `negligible flux` is a variable that is being tested.
Except for obvious special cases like 0<x<1.